base/abstractarray.jl

```diff
@@ -373,9 +373,16 @@ zero{T}(x::AbstractArray{T}) = fill!(similar(x), zero(T))
 # While the definitions for LinearFast are all simple enough to inline on their
 # own, LinearSlow's CartesianRange is more complicated and requires explicit
 # inlining.
-start(A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
-next(A::AbstractArray,i) = (@_inline_meta(); (idx, s) = next(i[1], i[2]); (A[idx], (i[1], s)))
-done(A::AbstractArray,i) = done(i[1], i[2])
+start(A::AbstractArray) = (@_inline_meta(); _start(linearindexing(A), A))
+_start(::LinearFast, A::AbstractArray) = 1
+_start(::LinearSlow, A::AbstractArray) = (@_inline_meta(); itr = eachindex(A); (itr, start(itr)))
```

Comment:
This concerns me a little bit — has there been a performance regression since https://github.com/JuliaLang/julia/pull/10704?  Or do reshaped arrays just exercise this differently in a way that makes this necessary?

It does make things a little more complicated to document and extend, but good performance trumps that.  Right now, you just need to define your own `eachindex` iterator, and iteration will always use it… even if you've defined your own custom `LinearIndexing` subtype.  If you change this method to `_start(::LinearIndexing, ::AbstractArray)`, all will be good.